WalletWasabi.Tests/UnitTests/SmartTransactionTests.cs

```diff
@@ -82,46 +82,245 @@ public void FirstSeenBackwardsCompatibility()
+			{
+				foreach (var inp in new[]
+				{
+					// Normal input.
+					$"{txHash}:{txHex}:{height}:{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}",
+					// Whitespaces.
+					$"     {txHash} :   {txHex}  :  {height}  :  {blockHash} :  {blockIndex}  : {label}    : {unixSeconds}     :    {isReplacement}  ",
+					// Don't fail on more inputs.
+					$"{txHash}:{txHex}:{height}:{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}::",
+					$"{txHash}:{txHex}:{height}:{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}:bar:buz",
+					// Can leave out some inputs.
+					$":{txHex}:{height}:{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}",
+					$"{txHex}:{height}:{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}",
+					$"{txHash}:{txHex}::{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}",
+					$"{txHash}:{txHex}:{height}::{blockIndex}:{label}:{unixSeconds}:{isReplacement}",
+					$"{txHash}:{txHex}:{height}:{blockHash}::{label}:{unixSeconds}:{isReplacement}",
+					$"{txHex}",
+				})
+				{
+					stx = SmartTransaction.FromLine(inp, net);
+
+					if (inp == $"{txHex}")
+					{
+						Assert.Equal(txHash, stx.GetHash().ToString());
+						Assert.Equal(txHex, stx.Transaction.ToHex());
+						Assert.Equal(Height.Unknown, stx.Height);
+						Assert.Null(stx.BlockHash);
+						Assert.Equal(0, stx.BlockIndex);
+						Assert.True(stx.Label.IsEmpty);
+						Assert.Equal(stx.FirstSeen.UtcDateTime, DateTime.UtcNow, TimeSpan.FromSeconds(1));
+						Assert.False(stx.IsReplacement);
+					}
+					else
+					{
+						Assert.Equal(txHash, stx.GetHash().ToString());
+						Assert.Equal(txHex, stx.Transaction.ToHex());
+						Assert.Equal(height, stx.Height.ToString());
+						Assert.Equal(blockHash, Guard.Correct(stx.BlockHash?.ToString()));
+						Assert.Equal(blockIndex, stx.BlockIndex.ToString());
+						Assert.Equal(label, stx.Label.ToString());
+						Assert.Equal(unixSeconds, stx.FirstSeen.ToUnixTimeSeconds().ToString());
+						Assert.Equal(isReplacement, stx.IsReplacement.ToString());
+					}
+				}
+			}
+
+			// Null and empty arguments.
+			string input = $"{txHash}:{txHex}:{height}:{blockHash}:{blockIndex}:{label}:{unixSeconds}:{isReplacement}";
+			Network network = null;
+			Assert.Throws<ArgumentNullException>(() => SmartTransaction.FromLine(input, network));
+			network = Network.Main;
+			input = "";
+			Assert.Throws<ArgumentException>(() => SmartTransaction.FromLine(input, network));
+			input = " ";
+			Assert.Throws<ArgumentException>(() => SmartTransaction.FromLine(input, network));
+			input = null;
+			Assert.Throws<ArgumentNullException>(() => SmartTransaction.FromLine(input, network));
+			input = "";
+			Assert.Throws<ArgumentException>(() => SmartTransaction.FromLine(input, network));
+			input = " ";
+			Assert.Throws<ArgumentException>(() => SmartTransaction.FromLine(input, network));
```

Comment:
Aren't the last two assert repeated?

Reply:
The second one has an space.

Reply:
No, I meant these two asserts exist already [here](https://github.com/zkSNACKs/WalletWasabi/pull/2216/files#diff-6f2c88d7e4f95ba03f31fc5c62b74431R210) and [here](https://github.com/zkSNACKs/WalletWasabi/pull/2216/files#diff-6f2c88d7e4f95ba03f31fc5c62b74431R212)

Reply:
Yes, they were duplicated. Removed.